backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/extraction/impl/LuceneCollectorsBuilder.java

```diff
@@ -52,33 +57,58 @@ public DistanceCollector addDistanceCollector(String absoluteFieldPath, GeoPoint
 	}
 
 	public LuceneCollectors build() {
-		Collector compositeCollector;
+		TopDocsCollector<?> topDocsCollector = null;
+		LuceneScoreDocPostProcessor scoreDocPostProcessor = null;
+		if ( requireTopDocs && maxDocs > 0 ) {
+			if ( sort == null ) {
+				topDocsCollector = TopScoreDocCollector.create(
+						maxDocs,
+						// TODO HSEARCH-3517 Avoid tracking the total hit count when possible
+						// Note this will also require to change how we combine collectors,
+						// as MultiCollector explicitly ignores the total hit count optimization
+						Integer.MAX_VALUE
+				);
+			}
+			else {
+				if ( requireScore ) {
+					// Since https://issues.apache.org/jira/browse/LUCENE-8412 (Lucene 8.0.0),
+					// TopFieldCollector returns TopDocs whose ScoreDocs do not contain a score...
+					// Thus we will have to get the scores ourselves.
+
+					// First possibility: if there's a SCORE sort field, use the value of that field
+					SortField[] sortFields = sort.getSort();
+					for ( int i = 0; i < sortFields.length; i++ ) {
+						SortField sortField = sortFields[i];
+						if ( sortField.getType() == SortField.Type.SCORE ) {
+							scoreDocPostProcessor = new LuceneTopFieldCollectorScoreSortPostProcessor( i );
```

Comment:
I think that this assignment will be overridden by the other at line 90.

Reply:
True; I'll fix that.